`peribolos`: add org roles feature

hoxhaeris opened the pull request:
# Add Organization Roles Support to Peribolos

This PR adds support for managing GitHub organization custom roles in Peribolos, enabling declarative configuration of role assignments to teams and users.

## Changes

### 1. Configuration Schema (`pkg/config/org/org.go`)
- Added `Roles` field to `Config` struct for declaring role assignments
- Implemented `ValidateRoles()` for pre-flight validation with case-insensitive team/user matching
- Uses `github.NormLogin()` for consistent username normalization

### 2. GitHub API Client (`pkg/github/client.go`)
Added methods for organization role management:
- `ListOrganizationRoles()`, `ListTeamsWithRole()`, `ListUsersWithRole()`
- `AssignOrganizationRoleToTeam()`, `RemoveOrganizationRoleFromTeam()`
- `AssignOrganizationRoleToUser()`, `RemoveOrganizationRoleFromUser()`

All mutating operations respect dry-run mode.

### 3. Main Implementation (`cmd/peribolos/main.go`)
- Added `--fix-org-roles` flag to enable role management (gatekeeper)
- Implemented `configureOrgRoles()` 
- Team name -> slug resolution for correct API calls
- Slug -> team name mapping in dump mode for idempotent configurations
- Roles configured after teams are fully populated

### 4. Test Coverage (`pkg/config/org/org_test.go`, `cmd/peribolos/main_test.go`)
- 12 validation test cases (nested teams, case-insensitivity, duplicates)
- 6 team assignment test cases
- 8 user assignment test cases

## Example Configuration

```yaml
orgs:
  my-org:
    teams:
      security-team:
        members: [user1, user2]
      compliance-team:
        members: [user3, user4]

    admins:
      - admin-user

    roles:
      security-manager:
        teams: [security-team]
        users: [admin-user]

      billing-manager:
        teams: [compliance-team]
        users: [admin-user]
```

## Usage

```bash
# Validate configuration only
peribolos --config-path=org.yaml

# Apply with dry-run
peribolos --config-path=org.yaml --fix-org-roles --confirm=false

# Apply changes
peribolos --config-path=org.yaml --fix-org-roles --confirm=true

# Dump current organization state
peribolos --dump my-org --github-token-path=token
```

hoxhaeris committed:
peribolos: add org roles feature

hoxhaeris commented:
/test pull-prow-unit-test

hoxhaeris commented:
/test pull-prow-unit-test

hoxhaeris commented:
/test pull-prow-verify-lint

hoxhaeris commented:
/unhold

ivankatliarchuk commented on pkg/github/client.go:
```diff
@@ -5164,3 +5176,163 @@ func (c *client) ListRepoInvitations(org, repo string) ([]CollaboratorRepoInvita
+	c.log("ListOrganizationRoles", org)
+	if c.fake {
+		return nil, nil
+	}
```
since ListOrganizationRoles returns nil, nil in fake mode, and configureOrgRoles early-exits when len(roles) == 0, any test that accidentally uses the real github.Client fake (rather than fakeOrgRolesClient) would pass trivially without exercising any role logic. This is a test-trap worth noting

hoxhaeris committed:
peribolos: fix org roles validation, error handling, and API alignment

- Move role-exists-in-GitHub validation before mutation loop
- Fix invitee fetching when only --fix-org-roles is set
- Remove phantom RoleID/RoleName/Type fields from OrganizationRoleAssignment
- Filter indirect user assignments and secret teams from dump output
- Propagate per-role errors in dump instead of silently continuing
- Detect case-insensitive role name collisions in ValidateRoles
- Add OrganizationRolesClient stubs to FakeClient
- Remove redundant dry-run guards in client methods
- Extract shared listRoleAssignments helper
- Decouple --fix-org-roles from --fix-teams requirement
- Aggregate team resolution errors instead of early return
- Replace tautological tests with real integration tests
- Add tests for error propagation, indirect filtering, pre-mutation validation

hoxhaeris commented on cmd/peribolos/main_test.go:
```diff
@@ -3970,3 +4059,683 @@ func TestConfigureCollaborators_PermissionMatrix_PendingInvitationUpdates(t *tes
 }
+
+// Test organization roles functionality - comprehensive tests
+func TestOrganizationRoles(t *testing.T) {
```
Removed

hoxhaeris commented:
/remove-lifecycle stale

hoxhaeris committed:
fix(peribolos): exclude unassigned roles from dump output

Only include organization roles in the dump when they have at least one
team or user assigned. Previously every predefined GitHub role appeared
in the output as an empty object, adding noise to diffs and configs.

hoxhaeris commented:
Rebased on latest main and added a small fix: the dump output now excludes organization roles that have no team or user assignments, so predefined GitHub roles no longer clutter the output as empty objects.

hoxhaeris commented:
/unhold
